### PR TITLE
Added segmenting support to REST API

### LINKED ISSUE #1035: Analytics: REST API Segmenting Support
Background: p7bje6-1bY-p2

This functionality has impact on Products, Variations, Coupons, Categories and Taxes Reports.

/cc @peterfabian if you would like to fill out some more details with your current thoughts to this issue.

### PR DESCRIPTION
**Pull request by peterfabian**
Fixes #1035.

First version of segmenting for `/orders/stats` (thus it should potentially also work for `/revenue/stats`, as that one uses the same data store).

Supports segmenting `orders/stats` by:
- product: `http://local.wordpress.test/wp-json/wc/v3/reports/orders/stats/?after=2018-01-11T12:52:01&before=2019-01-08T22:52:00&interval=month&order=desc&orderby=date&segmentby=product`
- variation (when one product is selected by a filter, e.g.: `product_includes=X`): `http://local.wordpress.test/wp-json/wc/v3/reports/orders/stats/?after=2018-01-11T12:52:01&before=2019-01-08T22:52:00&interval=month&order=desc&orderby=date&segmentby=variation&product_includes=9`
- category: `http://local.wordpress.test/wp-json/wc/v3/reports/orders/stats/?after=2018-01-11T12:52:01&before=2019-01-08T22:52:00&interval=month&order=desc&orderby=date&segmentby=category`
- coupon: `http://local.wordpress.test/wp-json/wc/v3/reports/orders/stats/?after=2018-01-11T12:52:01&before=2019-01-08T22:52:00&interval=month&order=desc&orderby=date&segmentby=coupon`
- customer_type: `http://local.wordpress.test/wp-json/wc/v3/reports/orders/stats/?after=2018-01-11T12:52:01&before=2019-01-08T22:52:00&interval=month&order=desc&orderby=date&segmentby=customer_type`

Segmenting of `products/stats` by:
- product
- category
- variation

Segmenting of `coupons/stats` by:
- product
- variation
- coupon
- category

Segmenting of `taxes/stats` by:
- tax_rate_id
- are other segments needed?

### Notes:
- At the moment, the paging and sorting only applies to intervals, not to segments (second dimension), i.e. it's adding all segments to totals and all intervals returned in the response. I'm not sure if that is a viable way to handle it, in case of 1000s of products or variations, it could be slow (?).
- The segments are currently always sorted by `segment_id`--does this need to change somehow?
- There is some creative math in some places, e.g. if shipping or coupons need to be split by product/variation, etc. Currently, 
  - shipping and shipping_tax are split in `wc_order_product_lookup` according to item quantity, e.g. order 1 has 2 line items: 4 x product_1 and 3 x product_2. Shipping for product_1 is thus 4/7 * total shipping, for product_2 it's 3/7 of total shipping for the whole order.
  - coupon amount for `wc_order_product_lookup` is calculated as `$order_item->get_subtotal( 'edit' ) - $order_item->get_total( 'edit' )`
  - refunds are split based on line item id and handled in `get_order_refund_items`
- I wasn't sure how the accounting should work in some cases, e.g. when segmenting /orders/stats by product, `net_revenue` and `num_items_sold` only include revenue/items for particular product, but `avg_items_per_order` and `avg_order_value` still refer to the whole order (in the sense that asking 'what is the average number of items per order for orders which contain product X' should probably include other products). I suggest creating a new issue/feature request, if this should be implemented differently. 
- I'm not sure all the edge cases are handled correctly, things to consider for /orders/stats (I plan to cover these by unit tests):
  - 2 records with the same product_id could appear in one order, e.g. when 2 variations of the same product are bought in the same order
  - 2 records with same product_id and variation_id could appear in one order, e.g. when hoodie variation is color=blue and size=any, then buying 2 sizes of blue hoodie will create this situation
  - multiple coupons can be assigned to one order
  - products can belong to multiple categories

### TODOs:
- [ ] update `orders/stats` to use new 'definition' of new/returning customer

### CODE REVIEW

File: includes/class-wc-admin-reports-coupons-stats-segmenting.php [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+				'order_level'   => $this->get_segment_selections_order_level( $table_name ),
+			);
+			$segmenting_from           = "INNER JOIN $product_segmenting_table ON ($table_name.order_id = $product_segmenting_table.order_id)";
+			$segmenting_groupby        = $product_segmenting_table . '.product_id';
+			$segmenting_dimension_name = 'product_id';
+
+			$segments = $this->get_product_related_segments( $type, $segmenting_selections, $segmenting_from, $segmenting_where, $segmenting_groupby, $segmenting_dimension_name, $table_name, $query_params, $unique_orders_table );
+		} elseif ( 'variation' === $this->query_args['segmentby'] ) {
+			if ( ! isset( $this->query_args['product_includes'] ) || count( $this->query_args['product_includes'] ) !== 1 ) {
+				throw new WC_REST_Exception( 'woocommerce_rest_invalid_segmenting_variation', __( 'product_includes parameter need to specify exactly one product when segmenting by variation.', 'wc-admin' ), 400 );
```
Comment (timmyc) [inline includes/class-wc-admin-reports-coupons-stats-segmenting.php]: Should this validation take place in the REST controller logic instead of here - maybe a validation callback on the `segmentby` query arg? Just not sure if this logic will ever be invoked outside of the context of a REST request.

File: includes/class-wc-admin-reports-orders-stats-segmenting.php [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+	 * @param string $segmenting_groupby GROUP BY part of segmenting SQL query.
+	 * @param string $table_name Name of SQL table which is the stats table for orders.
+	 * @param array  $intervals_query Array of SQL clauses for intervals query.
+	 *
+	 * @return array
+	 */
+	protected function get_order_related_intervals_segments( $segmenting_select, $segmenting_from, $segmenting_where, $segmenting_groupby, $table_name, $intervals_query ) {
+		global $wpdb;
+		$limit_parts      = explode( ',', $intervals_query['limit'] );
+		$orig_rowcount    = intval( $limit_parts[1] );
```
Comment (timmyc) [inline includes/class-wc-admin-reports-orders-stats-segmenting.php]: Do we need to validate $limit_parts has two elements here?

File: includes/class-wc-admin-reports-segmenting.php [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+				'segment_id' => $segment_id,
+				'subtotals'  => $segment_data,
+			);
+			$segment_result[ $segment_id ] = $segment_datum;
+		}
+
+		return $segment_result;
+	}
+
+	/**
```
Comment (timmyc) [inline includes/class-wc-admin-reports-segmenting.php]: Thanks for these comment blocks, it really helped me grok what was going on here.

File: includes/data-stores/class-wc-admin-reports-products-data-store.php [hunk 4]
```diff
@@ -337,36 +337,103 @@ public static function sync_order_products( $order_id ) {
 			$order_item_id     = $order_item->get_id();
 			$quantity_refunded = isset( $refunds[ $order_item_id ] ) ? $refunds[ $order_item_id ]['quantity'] : 0;
 			$amount_refunded   = isset( $refunds[ $order_item_id ] ) ? $refunds[ $order_item_id ]['subtotal'] : 0;
+			$product_qty       = $order_item->get_quantity( 'edit' ) - $quantity_refunded;
+
+			// Shipping amount based on woocommerce code in includes/admin/meta-boxes/views/html-order-item(s).php
+			// distributed simply based on number of line items.
```
Comment (timmyc) [inline includes/data-stores/class-wc-admin-reports-products-data-store.php]: Would it be better for this logic to live inside another util/class so it could be reused elsewhere - or allow for some test coverage?

**Review by timmyc — COMMENTED** [review]
😮Peter some serious props for architecting and developing this solution. The test case alone with all the requisite setup for the objects needed to do the testing was intense. I've tagged @claudiosanches for another set of ( more qualified ) eyes to check this out. I had a few minor comments in the code. I truly appreciate the detailed doc blocks and example usage within the code.

File: includes/class-wc-admin-reports-orders-stats-segmenting.php [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+	 */
+	protected function get_segment_selections_product_level( $products_table ) {
+		$columns_mapping = array(
+			'num_items_sold' => "SUM($products_table.product_qty) as num_items_sold",
+			'gross_revenue'  => "SUM($products_table.product_gross_revenue) AS gross_revenue",
+			'coupons'        => "SUM($products_table.coupon_amount) AS coupons",
+			'refunds'        => "SUM($products_table.refund_amount) AS refunds",
+			'taxes'          => "SUM($products_table.tax_amount) AS taxes",
+			'shipping'       => "SUM($products_table.shipping_amount) AS shipping",
+			// TODO: product_net_revenue should already have refunds subtracted, so it should not be here. Pls check.
```
Comment (claudiosanches) [inline includes/class-wc-admin-reports-orders-stats-segmenting.php]: Please use `@todo`, to make easy to find it later.

File: includes/class-wc-admin-reports-orders-stats-segmenting.php [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+
+		$product_segmenting_table = $wpdb->prefix . 'wc_order_product_lookup';
+		$unique_orders_table      = 'uniq_orders';
+		$segmenting_where         = '';
+
+		// Product, variation, and category are bound to product, so here product segmenting table is required,
+		// while coupon and customer are bound to order, so we don't need the extra JOIN for those.
+		// This also means that segment selections need to be calculated differently.
+		if ( 'product' === $this->query_args['segmentby'] ) {
+			// TODO: how to handle shipping taxes when grouped by product?
```
Comment (claudiosanches) [inline includes/class-wc-admin-reports-orders-stats-segmenting.php]: Another `@todo` that needs to be updated.

File: includes/class-wc-admin-reports-segmenting.php [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+			);
+		} elseif ( 'category' === $this->query_args['segmentby'] ) {
+			$categories = get_categories(
+				array(
+					'taxonomy' => 'product_cat',
+				)
+			);
+			$segments   = wp_list_pluck( $categories, 'cat_ID' );
+		} elseif ( 'coupon' === $this->query_args['segmentby'] ) {
+			// TODO: switch to a non-direct-SQL way to get all coupons?
```
Comment (claudiosanches) [inline includes/class-wc-admin-reports-segmenting.php]: Another `@todo`

File: includes/data-stores/class-wc-admin-reports-products-data-store.php (hunk 4 above)
Comment (claudiosanches) [reply includes/data-stores/class-wc-admin-reports-products-data-store.php]: I think that should be great a new class or function, so we can use in both places, also we can include unit tests.

File: includes/data-stores/class-wc-admin-reports-products-data-store.php [hunk 8]
```diff
@@ -337,36 +337,103 @@ public static function sync_order_products( $order_id ) {
...
+				$total_shipping_amount = $order->get_shipping_total() - $refunded;
+			} else {
+				$total_shipping_amount = $order->get_shipping_total();
+			}
+			$shipping_amount = $total_shipping_amount / $order_items * $product_qty;
+
+			// Shipping amount tax based on woocommerce code in includes/admin/meta-boxes/views/html-order-item(s).php
+			// distribute simply based on number of line items.
+			$shipping_tax_amount = 0;
+			// TODO: if WC is currently not tax enabled, but it was before (or vice versa), would this work correctly?
```
Comment (claudiosanches) [inline includes/data-stores/class-wc-admin-reports-products-data-store.php]: One more `@todo`.

**Comment by claudiosanches** [comment]
I just read all the code and looks good, just see that need update `TODO` to `@todo`, so we can use any tool to find it later, of even searching in the editor makes hard if we use multiple formats.
Now going to run this code.

**Comment by peterfabian** [comment]
Work on segmenting for `taxes/stats` depends on #1394.

File: includes/api/class-wc-admin-rest-reports-coupons-stats-controller.php [hunk 9]
```diff
@@ -156,6 +157,36 @@ public function get_item_schema() {
...
+							'context'     => array( 'view', 'edit' ),
+							'readonly'    => true,
+							'properties'  => $data_values,
+						),
+					),
+				),
+			),
+		);
+
+		$totals = array_merge( $data_values, $segments );
```
[the remark this answers is not in the document]
Comment (peterfabian) [reply includes/api/class-wc-admin-rest-reports-coupons-stats-controller.php]: But $segments uses $data_values, I would prefer not to create an array that repeats parts of it multiple times, unless necessary.

File: includes/class-wc-admin-reports-coupons-stats-segmenting.php (hunk 1 above)
Comment (peterfabian) [reply includes/class-wc-admin-reports-coupons-stats-segmenting.php]: Updated exception handling in https://github.com/woocommerce/wc-admin/pull/1286/commits/42fff45f447541f4569a6a7cf3f3dcbad42ef7ba.
As I believe it does not make much sense to compare variations from different products, I kept the condition, at least for now.

File: includes/class-wc-admin-reports-orders-stats-segmenting.php (hunk 2 above)
Comment (peterfabian) [reply includes/class-wc-admin-reports-orders-stats-segmenting.php]: Yup, good idea, thanks. Updated in https://github.com/woocommerce/wc-admin/pull/1286/commits/c83f9b0015474cf503276779af9730620c64d41d.

File: includes/data-stores/class-wc-admin-reports-products-data-store.php (hunk 4 above)
Comment (peterfabian) [reply includes/data-stores/class-wc-admin-reports-products-data-store.php]: Factored the logic out to a class extending `WC_Order` in https://github.com/woocommerce/wc-admin/pull/1286/commits/f1dddc9243db0cc6808d9a13ecd4beb350d40ae4. 
Would be great to later add unit tests for it if time allows.

File: includes/api/class-wc-admin-rest-reports-coupons-stats-controller.php [hunk 10]
```diff
@@ -302,6 +333,17 @@ public function get_collection_params() {
 				'type' => 'integer',
 			),
 		);
+		$params['segmentby']  = array(
+			'description'       => __( 'Segment the response by additional constraint.', 'wc-admin' ),
+			'type'              => 'string',
+			'enum'              => array(
+				'product',
+				'variation',
+				'category', // will we support this?
```
[the remark this answers is not in the document]
Comment (peterfabian) [reply includes/api/class-wc-admin-rest-reports-coupons-stats-controller.php]: Updated all the `// TODO`s in the wc-admin, so hopefully all are fixed now (https://github.com/woocommerce/wc-admin/pull/1286/commits/fed469ebb2c24ec987a35cb449bb85dcb7a848f8).

**Comment by timmyc** [comment]
Is this ready for another review @peterfabian ?

**Comment by peterfabian** [comment]
Yes please! @timmyc

**Review by timmyc — APPROVED** [review]
@peterfabian I think we should get this in once it gets rebased - even though Github is saying it can be merged in okay, when attempt to rebase locally it seems to have a few conflicts.

**Review by emdashcodes — APPROVED** [review]
I have a few minor comments related to the schema, but otherwise this is looking really good to me! All `phpunit` tests pass. Pre-approving pending those minor changes.

I've also been testing product/variation segmentation heavily with #1391 which has been working great.

**Comment by peterfabian** [comment]
Thanks @justinshreve . Should be solved by https://github.com/woocommerce/wc-admin/pull/1286/commits/994c9b89752419b013e3528bc6f481f0c97f1893.

**Comment by justinshreve** [comment]
> As I prefer merging over rebasing, so I merged master into this branch whenever a conflict appeared. Last time master was merged is 3 days ago. If you think it's a good idea, I can rebase it, but I would prefer not to, unless necessary, as I used merging throughout the life of the branch.

I think the conflicts Timmy saw are from earlier commits (since `git rebase master` applies commits one at a time) -- so I agree it would be more trouble then it is worth to rebase this.

Since GitHub here is clean, I would say go ahead and merge 👍.

**Review by claudiosanches — APPROVED** [review]
Great work!
